components/odh-notebook-controller/e2e/notebook_creation_test.go

```diff
@@ -0,0 +1,243 @@
+			}
+		}
+	}
+}
+
+func (tc *testContext) testRouteCreation(t *testing.T) {
+	nbRoute, err := tc.getNotebookRoute()
+	require.NoErrorf(t, err, "error getting Notebook Route :%s", testNotebookName)
+
+	isReady := false
+	for _, ingress := range nbRoute.Status.Ingress {
+		if strings.Contains(ingress.Host, testNotebookName) {
+			for _, condition := range ingress.Conditions {
+				if condition.Type == routev1.RouteAdmitted && condition.Status == v1.ConditionTrue {
+					isReady = true
+				}
+			}
+		}
+	}
+	require.EqualValuesf(t, true, isReady, "Notebook Route %s is not Ready.", testNotebookName)
+}
+
+func (tc *testContext) testNotebookTraffic(t *testing.T) {
+	resp, err := tc.curlNotebookEndpoint()
+	require.NoErrorf(t, err, "error accessing Notebook Endpoint")
+	require.EqualValuesf(t, 200, resp.StatusCode, "Unexpected response from Notebook Endpoint")
+}
+
+func (tc *testContext) testNotebookCulling(t *testing.T) {
+	// Create Configmap with culling configuration
+	cullingConfigMap := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "notebook-controller-culler-config",
+			Namespace: tc.testNamespace,
+		},
+		Data: map[string]string{
+			"ENABLE_CULLING":        "true",
+			"CULL_IDLE_TIME":        "2",
+			"IDLENESS_CHECK_PERIOD": "1",
+		},
+	}
+
+	_, err := tc.kubeClient.CoreV1().ConfigMaps(tc.testNamespace).Create(context.TODO(), cullingConfigMap,
+		metav1.CreateOptions{})
+	require.NoErrorf(t, err, "error creating configmap: notebook-controller-culler-config")
+	// Restart the deployment to get changes from configmap
+	controllerDeployment, err := tc.kubeClient.AppsV1().Deployments(tc.testNamespace).Get(context.TODO(),
+		"notebook-controller-deployment", metav1.GetOptions{})
+	require.NoError(t, err)
+	// Scale deployment to 0
+	expectedReplicas := int32(0)
+	controllerDeployment.Spec.Replicas = &expectedReplicas
+	controllerDeployment, err = tc.kubeClient.AppsV1().Deployments(tc.testNamespace).Update(context.TODO(),
+		controllerDeployment, metav1.UpdateOptions{})
+	require.NoError(t, err)
+	// Scale deployment to 1
+	expectedReplicas = int32(1)
+	controllerDeployment.Spec.Replicas = &expectedReplicas
+	_, err = tc.kubeClient.AppsV1().Deployments(tc.testNamespace).Update(context.TODO(),
+		controllerDeployment, metav1.UpdateOptions{})
+	require.NoError(t, err)
+
+	// Wait for server to shut down after 'CULL_IDLE_TIME' minutes(around 2.5 minutes)
+	time.Sleep(150 * time.Second)
+	// Verify that the notebook kernel has shutdown, and the notebook endpoint returns 503
+	resp, _ := tc.curlNotebookEndpoint()
+	require.EqualValuesf(t, 503, resp.StatusCode, "Unexpected response from Notebook Endpoint")
+
+	// Delete the culling configuration Configmap once the test is completed
+	err = tc.kubeClient.CoreV1().ConfigMaps(tc.testNamespace).Delete(context.TODO(),
+		cullingConfigMap.Name, metav1.DeleteOptions{})
+	require.NoErrorf(t, err, "error creating configmap: notebook-controller-culler-config")
```

Comment:
Move this to a defer function to delete the configmap and restart the controller again to revert the culling configuration when the test is done.

https://gobyexample.com/defer